Track Stakepool Retirements in the DB

jonathanknowles opened the pull request:
# Related Issues

#1815 
#1816 
#1817 
#1819 
#1880 

# Overview

This PR:

- [x] Extends the pool DB schema to make it possible to store:
    - [x] retirement certificates.
    - [x] the _order_ of publication of certificates _within the same block_, for both registration and retirement certificates. This is necessary, as the intra-block order is _significant_.<br><br>

- [x] Adds the `PoolLifeCycleStatus` type, which indicates the current lifecycle stage of a pool.

    There are currently _three_ possibilities:
    1. `NotRegistered`
        (indicates that a pool has never been registered)
    2. `Registered`
        (indicates that a pool has been registered but _not_ retired)
        (provides a registration certificate)
    3. `RegisteredAndRetired`
        (indicates that a pool has been registered _and_ retired)
        (provides both a registration and a retirement certificate)<br><br>

- [x] Adds the `readPoolLifeCycleStatus` function to the pool DB layer, making it possible to determine the current lifecycle status of a pool. 

- [x] Updates the `retirement` field of `ApiStakePool` with live information.

# Property Tests

This PR adds property tests to ensure that:

- [x] the `determinePoolLifeCycleStatus` function respects the correct order of precedence for registration and retirement certificates:
    - for a given pool, a registration certificate always _supercedes_ a prior retirement certificate.
    - for a given pool, a retirement certificate always _augments_ the latest registration certificate.
- [x] `readPoolLifeCycleStatus` queries work correctly in the presence of blocks containing multiple certificates for the same pool.
- [x] database rollback works correctly in the presence of retirement certificates.

# Future Work

A **_future_** PR will add integration tests to ensure that:

- retiring a pool eventually results in a correct update to the `retirement` field of `ApiStakePool`, when read through the `ListStakePools` operation.
- retiring and then re-registering a pool clears the `retirement` field. 

See QA section of https://github.com/input-output-hk/cardano-wallet/issues/1819.

jonathanknowles committed:
Add `PoolRetirement` SQLite table.

jonathanknowles committed:
Reformat code to comply with coding standards.

jonathanknowles committed:
Add {put,read}PoolRetirement DB operations.

jonathanknowles committed:
Call `putPoolRetirement` on encountering a retirement certificate.

jonathanknowles committed:
Introduce type `SlotInternalIndex`.

jonathanknowles committed:
Make `SlotInternalIndex` values persistable.

jonathanknowles committed:
Simplify function `putPoolRegistration`.

jonathanknowles committed:
Add `slot_internal_index` fields.

Add `slot_internal_index` fields to DB pool certificate tables.

jonathanknowles committed:
Record the order in which certificates appear.

In the database, record the order in which pool certificates appear in a
Shelley block.

jonathanknowles committed:
Introduce type `PoolRegistrationStatus`.

Introduce type `PoolRegistrationStatus` and related functions.

Type `PoolRegistrationStatus` indicates the registration status of a
pool, which can be one of the following options:

  1. Not registered.
  2. Registered.
  3. Registered but retired.

In addition, this change introduces the `readPoolRegistrationStatus`
function, which read the pool registration status of a given pool on
any given database backend.

jonathanknowles committed:
Return the real retirement status in ApiStakePool.

Return the real retirement status of a Shelley stake pool in ApiStakePool.

jonathanknowles committed:
Rename stake-pool-related data types.

Use the following names:

- PoolLsqData
- PoolDbData

Justification: the data represented by these types is no longer just
restricted to metrics.

jonathanknowles committed:
Rename stake-pool-related record fields.

jonathanknowles committed:
Introduce `CertificatePublicationTime` type.

The `CertificatePublicationTime` type represents an abstract notion of a
certificate publication time.

Certificates published at later times take precedence over certificates
published at earlier times.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -180,22 +192,60 @@ mReadStakeDistribution epoch db@PoolDatabase{distributions} =
+    :: CertificatePublicationTime
+    -> PoolRegistrationCertificate
+    -> ModelPoolOp ()
+mPutPoolRegistration sp registration db =
```
Given that the type is `CertificatePublicationTime`, we should probably also rename `sp` to `cpt` for consistency.

```suggestion
mPutPoolRegistration cpt registration db =
```

jonathanknowles replied:
Fixed in a75c879aabde68edcc1389a38a9e09b0bbdf17a7.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -180,22 +192,60 @@ mReadStakeDistribution epoch db@PoolDatabase{distributions} =
+    :: CertificatePublicationTime
+    -> PoolRetirementCertificate
+    -> ModelPoolOp ()
+mPutPoolRetirement sp retirement db =
```
Given that the type is `CertificatePublicationTime`, we should probably also rename sp to cpt for consistency.

```suggestion
mPutPoolRetirement cpt retirement db =
```

jonathanknowles replied:
Fixed in a75c879aabde68edcc1389a38a9e09b0bbdf17a7.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite/TH.hs:
```diff
@@ -81,21 +81,31 @@ PoolOwner sql=pool_owner
     poolOwnerIndex      Word8        sql=pool_owner_index
 
     Primary poolOwnerPoolId poolOwnerSlot poolOwnerOwner poolOwnerIndex
-    Foreign PoolRegistration fk_registration_pool_id poolOwnerPoolId poolOwnerSlot ! ON DELETE CASCADE
```
We should replace this with suitable foreign key constraints.

jonathanknowles replied:
> We should replace this with suitable foreign key constraints.

Fixed in b28504ec32370917e83bd4b593b1cf7c45c76cfc.

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -315,6 +316,22 @@ instance PathPiece SlotId where
+    toJSON = toJSON . unSlotInternalIndex
+
+instance FromJSON SlotInternalIndex where
+    parseJSON = fmap SlotInternalIndex . parseJSON
```
We should add round trip serialization tests for theses instances.

jonathanknowles replied:
> We should add round trip serialization tests for theses instances.

Fixed in cbab6b40db6d770a5dc2573969b142c44f3987c0.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +187,64 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+-- Certificates published at later times take precedence over certificates
+-- published at earlier times.
+--
+type CertificatePublicationTime = (SlotId, SlotInternalIndex)
```
Consider making this into a `newtype`.

rvl replied:
It would need to be a `data` because there are two values.

I think remove `newtype SlotInternalIndex` and make this a data type.

Can you find a less wordy name for it?

jonathanknowles replied:
I've promoted this type synonym into a concrete `data` type. (See 01fc73d8193597645069f1d1a4a5d7c0e79eeae2.)

jonathanknowles replied:
@rvl wrote:
> Can you find a less wordy name for it?

I can't (currently) think of a shorter name for this.

For example, we could use `SlotIndex`, but IMO that doesn't make it clear enough that the index is _internal_ to a slot.

Do you have any ideas for a better name?

jonathanknowles replied:
I've removed the `SlotInternalIndex` type, inlining the definition into `CertificatePublicationTime`. (See c7e5da754889c5ba684fbd82fa3d899996ff0624.)

jonathanknowles committed:
Use identifier names that are consistent with types.

jonathanknowles committed:
Add error checking to `determinePoolRegistrationStatus`.

jonathanknowles committed:
Add `Arbitrary` instance for `PoolRetirementCertificate`.

jonathanknowles committed:
Use OverloadedLabels extension to avoid ambiguity.

The field `poolId` appears in multiple record types, including:

- PoolRegistrationCertificate
- PoolRetirementCertificate

jonathanknowles committed:
Generalize the type of time in `determinePoolRegistrationStatus`.

This makes it easier to write property tests, as we are free to use the
type of our choice to represent time.

jonathanknowles committed:
Add property `prop_determinePoolRegistrationStatus_orderCorrect`.

jonathanknowles committed:
Add property `prop_determinePoolRegistrationStatus_neverRegistered`.

jonathanknowles committed:
Add property `prop_determinePoolRegistrationStatus_differentPools`.

jonathanknowles committed:
Extend `prop_poolRegistration` to verify certificate publication times.

jonathanknowles committed:
Add property `prop_poolRetirement`.

jonathanknowles committed:
Add `Arbitrary` instance for `PoolCertificate`.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -354,22 +389,39 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+            forM_ (certificateIndices `zip` certificates) $ \case
+                (slotInternalIndex, Registration cert) -> do
+                    liftIO $ traceWith tr $ MsgStakePoolRegistration cert
+                    putPoolRegistration (slot, slotInternalIndex) cert
```
isn't this slot internal index superfluous? 

a) All certificates within a block are inserted _atomically_, so they're either all inserted, or none of them are. 
b) A certificate present earlier in a block will be overridden by one later in the block
c) This is a strategy we've been using so far for all certificates, so I find the discrepancy here rather confusing. It suggests there's something special about these certificates, although there isn't (or am I missing something?) 

I'd suggest to drop the `slotInternalIndex` to not add on top of the existing complexity.

jonathanknowles replied:
> isn't this slot internal index superfluous?

It turns out that:

1. A _single block_ can contain multiple certificates relating to the same pool.
2. A _single transaction_ can contain multiple certificates relating to the same pool.
3. The _relative order_ of these certificates is _significant_.
4. In general, earlier certificates are made obsolete by later certificates, but in particular:
    - a registration certificate will always _supersede_ an earlier retirement certificate (making that earlier retirement certificate obsolete);
    - a retirement certificate will always _augment_ the latest known registration certificate.

See [this thread](https://input-output-rnd.slack.com/archives/CCRB7BU8Y/p1593584824393100) for more information.

After discussing this with @rvl and @Anviking, we identified two potential methods of dealing with this issue:
- **Option A**: when processing a given block in a given slot **_s_**, if there are multiple registration and/or retirement certificates for a given pool **_p_**, then only record the latest registration and retirement certificates for that pool in the database, and ignore all prior certificates (subject to the rules stated above).
- **Option B**: when processing a given block in a given slot **_s_**, always record _all_ certificates for a given pool **_p_**, but also record the _order_ in which they appeared within the block. When processing a query for a given pool, fetch the latest registration and retirement certificates, and at _that point_ determine their relative precedence.

Comparing these two options:
* Option A has the advantage of retaining a simpler database design.
* Option B has the advantage of retaining more information to aid with debugging.

On balance, we felt that option B is the better choice. While it is true that this option requires more invasive changes to the pool database (an extra column in three tables to record the intra-slot order), in the event we need to perform debugging, the extra information (a record of all certificates that we've encountered, rather than just a subset) could prove to be very helpful.

jonathanknowles committed:
Delete retirement certificates when cleaning the SQLite database.

jonathanknowles committed:
Add property `prop_readPoolRegistrationStatus`.

jonathanknowles committed:
Add `slot_internal_index` field to `PoolOwner` table.

jonathanknowles committed:
Add property `prop_multiple_putPoolRegistration_single_readPoolRegistration`.

jonathanknowles committed:
Add property `prop_multiple_putPoolRetirement_single_readPoolRetirement`.

jonathanknowles committed:
Add further explanatory comments to properties.

Add further explanatory comments to properties in
`Cardano.Pool.DB.Properties`.

jonathanknowles committed:
Verify persistence of `SlotInternalIndex` values.

Verify that values of type `SlotInternalIndex` can be successfully
written to and read from the database without loss of information.

jonathanknowles committed:
Remove retirement certificates during database rollback.

jonathanknowles committed:
Add property `prop_rollbackRetirement`.

jonathanknowles committed:
Reinstate foreign key constraint for `PoolOwner` table.

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -221,16 +230,24 @@ combineDbAndLsqData =
           -- For the moment, we always report that a pool will never retire.
           -- See https://github.com/input-output-hk/cardano-wallet/milestone/89
-        , Api.retirement = Nothing
+        , Api.retirement =
```
Is the TODO comment above still valid?

jonathanknowles replied:
> Is the TODO comment above still valid?

You're right! It's redundant.

Fixed in 1ab38cd965f5d12b210aa841856dc50797de2a12.

rvl commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +190,98 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+    => Maybe (certificatePublicationTime, PoolRegistrationCertificate)
+    -> Maybe (certificatePublicationTime, PoolRetirementCertificate)
+    -> PoolRegistrationStatus
+determinePoolRegistrationStatus = f
```
I'm pretty sure this could be made shorter with `case` and guards, without the need for helper functions f and g.
```suggestion
determinePoolRegistrationStatus a b = case (a, b) of
```

jonathanknowles replied:
> I'm pretty sure this could be made shorter with `case` and guards, without the need for helper functions f and g.

It does indeed save a few lines (8 lines added in return for 12 lines deleted).

I've fixed this in c11624eb76afacf8448f279656da74b940f8b443.

rvl commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +190,98 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+-- Use the 'readPoolRegistrationStatus' function to query the registration
+-- status for a particular pool and database backend.
+--
+data PoolRegistrationStatus
```
Why not represent this status as `Maybe (Registration, Maybe Retirement)`? I find the standard types easier to understand.

jonathanknowles replied:
> Why not represent this status as `Maybe (Registration, Maybe Retirement)`? I find the standard types easier to understand.

Having a dedicated type means we have a single, obvious place to document what each of the states means. A single source of truth, if you like.

Consider the following pair of functions, both of which return the same type:

```hs
determinePoolLifeCycleStatus
    :: Ord publicationTime
    => Maybe (publicationTime, PoolRegistrationCertificate)
    -> Maybe (publicationTime, PoolRetirementCertificate)
    -> PoolLifeCycleStatus
```
```hs
readLifeCycleStatus
    :: DBLayer m
    -> PoolId
    -> m PoolLifeCycleStatus
```
A reader who is curious about what `PoolLifeCycleStatus` means can simply navigate to the definition and read the documentation for each case:

```hs
data PoolLifeCycleStatus
    = PoolNotRegistered
        -- ^ Indicates that a pool is not registered.
    | PoolRegistered
        PoolRegistrationCertificate
        -- ^ Indicates that a pool is registered BUT NOT marked for retirement.
        -- Records the latest registration certificate.
    | PoolRegisteredAndRetired
        PoolRegistrationCertificate
        PoolRetirementCertificate
        -- ^ Indicates that a pool is registered AND ALSO marked for retirement.
        -- Records the latest registration and retirement certificates.
```

IMO, even though the type `Maybe (PoolRegistrationCertificate, Maybe PoolRetirementCertificate)` is isomorphic to `PoolLifeCycleStatus`, it has the disadvantage of not being immediately clear to the uninformed reader what is meant by each of the cases.

We could document the meaning of each of the cases in all functions that consume or return this type, but that seems to be worse than just providing a dedicated type with its own documentation.

rvl commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -130,8 +158,20 @@ properties = do
             (property . prop_putStakeReadStake)
         it "putPoolRegistration then readPoolRegistration yields expected result"
             (property . prop_poolRegistration)
+        it "putPoolRetirement then readPoolRetirement yields expected result"
```
I can't seem to find any modifications to the DB state machine tests here.

I think you will get more thorough testing for your effort by updating the state machine generators and adding a few tags to check that they work.

jonathanknowles replied:
> I can't seem to find any modifications to the DB state machine tests here.
> 
> I think you will get more thorough testing for your effort by updating the state machine generators and adding a few tags to check that they work.

Definitely agree with this.

Unfortunately there doesn't seem to be a state machine test suite for the pool database.

Perhaps we can add one in a future PR? I've started a discussion in the team channel here: https://input-output-rnd.slack.com/archives/GBT05825V/p1594183858000500

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -172,27 +179,29 @@ newStakePoolLayer gp nl db = StakePoolLayer
-data PoolLsqMetrics = PoolLsqMetrics
+-- | Stake pool-related data that has been read from the node using a local
+--   state query.
+data PoolLsqData = PoolLsqData
```
Since you're changing the name anyway... I don't think `Lsq` is descriptive enough of what the data is. It only describes the mechanism used to fetch it. How about `PoolStateFromLedger`?

jonathanknowles replied:
> Since you're changing the name anyway... I don't think `Lsq` is descriptive enough of what the data is. It only describes the mechanism used to fetch it. How about `PoolStateFromLedger`?

I do tend to agree with you here.

But on the other hand, the data for `PoolLsqData` comes (in part) from another data type called `NodePoolLsqData` (defined in `Cardano.Wallet.Shelley.Network`). See the function `combineLsqData` for more info.

If we were to change this name, I feel that we should also change the names of `NodePoolLsqData` and `combineLsqData` for consistency, as these manipulate the same sets of data.

But perhaps that is outside the scope of this PR? We could always make another PR with these changes (to avoid making this PR too big).

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -293,26 +312,42 @@ combineChainData =
+    certificatesFromRegistrationStatus
+        :: PoolRegistrationStatus
+        -> Maybe (PoolRegistrationCertificate, Maybe PoolRetirementCertificate)
+    certificatesFromRegistrationStatus = \case
```
If PoolRegistrationStatus were replaced with the double Maybe then this function would not be needed.

jonathanknowles replied:
> If PoolRegistrationStatus were replaced with the double Maybe then this function would not be needed.

That's true. But IMO this would reduce the readability of the following functions, both of which use the `PoolRegistrationStatus` type.

* `determinePoolRegistrationStatus`
* `readPoolRegistrationStatus`

See my comment here for a justification: https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451304532

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -354,22 +389,39 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+            -- than certificates that appear earlier.
+            --
+            let certificateIndices = [minBound ..]
+            forM_ (certificateIndices `zip` certificates) $ \case
```
Can we please just say `zip [0..] certificates`.

jonathanknowles replied:
> Can we please just say `zip [0..] certificates`.

So we can't use `0` here, as `SlotInternalIndex` doesn't implement `Num`. But it does implement `Bounded`, hence the use of `minBound`.

As for using the name `certificateIndices`: I'm not sure we should get rid of that, as having the name serves as a kind of documentation for what these indices are supposed to be.

Perhaps there is a better name though?

jonathanknowles replied:
@rvl I've rewritten this part of the code in 01fc73d8193597645069f1d1a4a5d7c0e79eeae2.

Let me know if you think it could be revised further!

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -354,22 +389,39 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+            --
+            -- Certificates that appear later in a block have higher precedence
+            -- than certificates that appear earlier.
+            --
```
```suggestion
            -- When there are multiple certificates for a pool in one block,
            -- certificates that appear later in the block have higher precedence.
            -- It is possible to have a pool (re-)registration and retirement within the same block.
```

jonathanknowles replied:
@rvl I've revised this comment a bit to make it more useful. (See 9efad1d76dc39641e79ed8caeff7f7003493644e)

Let me know if you think it could be revised further. :+1:

rvl commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +190,98 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+        | regPoolId /= retPoolId =
+            differentPoolsError
+        | regTime > retTime =
+            PoolRegistered regCert
```
Is this case correct?
I think this is the only non-obvious about this function, and it's worth a comment.

jonathanknowles replied:
> Is this case correct?

Yes, it's correct. This case represents a re-registration (a registration that occurs after the prior publication of a retirement certificate), in which case the prior retirement is deemed to be obsolete. (See "pool inference rule" from the [Shelley ledger specification](https://hydra.iohk.io/build/3202141/download/1/ledger-spec.pdf))

> I think this is the only non-obvious about this function, and it's worth a comment.

Great point. I've added more comments in fcdfde6ce27c6f7744ecf7728ff1151908d4c959, which also adds a link to the specification.

rvl reviewed (changes requested):
Looks pretty good but I have a few niggles...

jonathanknowles committed:
Mark task to return real retirement status in API as complete.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451245130

jonathanknowles committed:
Simplify `determinePoolRegistrationStatus` pattern matching.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451247533

jonathanknowles committed:
Revise comment within `monitorStakePools`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451254738

jonathanknowles committed:
Document different cases in `determinePoolRegistrationStatus`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451255425

jonathanknowles committed:
Promote `CertificatePublicationTime` into a concrete data type.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r449738019

jonathanknowles committed:
Move types to `Cardano.Wallet.Primitive.Types`.

This change moves the following types to `Cardano.Wallet.Primitive.Types`:

- `CertificatePublicationTime`
- `PoolRegistrationStatus`

rvl commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +188,87 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+        | regPoolId /= retPoolId ->
+            -- Comparing certificates from different pools is a programming
+            -- error.
+            differentPoolsError
```
You may as well remove the comment and inline the differentPoolsError string -- it's self-documenting.

jonathanknowles replied:
> You may as well remove the comment and inline the differentPoolsError string -- it's self-documenting.

Thanks! I've removed the comment, and revised the error string to make it clear that it's a programming error.

(The only thing I haven't changed is to inline `differentPoolsError`, in order to keep the pattern matching fairly compact.)

rvl commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +188,87 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+            -- If a registration certificate and a retirement certificate
+            -- for the same pool appear to have been published at exactly
+            -- the same time, this indicates a programming error.
+            timeCollisionError
```
Same here - also the error messages should mention "programming error", so that if it ever occurs - which it won't - the user knows that it's not their fault.

jonathanknowles replied:
> Same here - also the error messages should mention "programming error", so that if it ever occurs - which it won't - the user knows that it's not their fault.

> You may as well remove the comment and inline the differentPoolsError string -- it's self-documenting.

Thanks! I've removed the comment, and revised the error string to make it clear that it's a programming error.

rvl commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -196,62 +200,114 @@ newDBLayer trace fp = do
+                    ]
+            selectFirst filterBy orderBy >>= \case
+                Nothing -> pure Nothing
+                Just meta -> do
```
How about this way?

```suggestion
            res <- selectFirst [ PoolRetirementPoolId ==. poolId ]
                    [ Desc PoolRetirementSlot
                    , Desc PoolRetirementSlotInternalIndex
                    ]
            forM res $ \meta -> do
```

jonathanknowles replied:
> How about this way?

Nice!

The function `readPoolRegistration` was originally written this way, so I extended the style when writing `readPoolRetirement`. But your way is shorter and clearer.

I've applied it to the pair of functions `readPool{Registration,Retirement}`

Fixed in: c79a3300883dfc4bd62c18ec23aa730bb4a208b6.

rvl commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -1771,6 +1784,37 @@ instance Buildable PoolRetirementCertificate where
+    { slotId
+        :: SlotId
+    , slotInternalIndex
+        :: SlotInternalIndex
```
Adding this newtype only clouds the meaning. Almost any integral type would be better.

jonathanknowles replied:
> Adding this newtype only clouds the meaning. Almost any integral type would be better.

Fair enough!

I've removed the `SlotInternalIndex` type by inlining it into `CertificatePublicationTime`, resulting in the following type:
```hs
data CertificatePublicationTime = CertificatePublicationTime
    { slotId
        :: SlotId
    , slotInternalIndex
        :: Word64
        -- ^ Indicates the relative position of a publication within a slot.
    }
    deriving (Eq, Generic, Ord, Show)
```
Fixed in c7e5da754889c5ba684fbd82fa3d899996ff0624.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +188,87 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+            , " pool id of retirement certificate: "
+            , show retPoolId
+            , " publication time of registration certificate: "
+            , show regPoolId
```
This should be a publication time, rather than a pool ID.

jonathanknowles replied:
Fixed in eae043b6cb8510cb292cb23aec211d3296edfd20.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -166,6 +188,87 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+            , " publication time of registration certificate: "
+            , show regPoolId
+            , " publication time of retirement certificate: "
+            , show retPoolId
```
This should be a publication time, rather than a pool ID.

jonathanknowles replied:
Fixed in eae043b6cb8510cb292cb23aec211d3296edfd20.

jonathanknowles committed:
Revise comments and errors in `determinePoolRegistrationStatus`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847/files#r451921232
https://github.com/input-output-hk/cardano-wallet/pull/1847/files#r451921583

jonathanknowles committed:
Show certificate publication times in error messages.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451932921
https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451932976

jonathanknowles committed:
Simplify definitions of `readPool{Registration,Retirement}`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451926351

jonathanknowles committed:
Inline `SlotInternalId` type into `CertificatePublicationTime`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r451927736

jonathanknowles committed:
Be more specific when deleting entries from the `PoolOwner` table.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1847#discussion_r452085881

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -196,62 +203,104 @@ newDBLayer trace fp = do
+        , putPoolRegistration = \cpt cert -> do
+            let CertificatePublicationTime {slotId, slotInternalIndex} = cpt
+            let poolId = view #poolId cert
+            deleteWhere [PoolOwnerPoolId ==. poolId, PoolOwnerSlot ==. slotId]
```
This looks as though it may be deleting _too many_ rows.

We probably want to restrict this to deleting rows with a matching value of `PoolOwnerSlotInternalIndex`.

jonathanknowles replied:
> This looks as though it may be deleting _too many_ rows.
> 
> We probably want to restrict this to deleting rows with a matching value of `PoolOwnerSlotInternalIndex`.

Fixed in 31f8e0fd8a15814b4624dde566c608de0fb44ec7.

jonathanknowles committed:
Use more precise sort order when listing pools.

jonathanknowles committed:
Don't attempt to execute nested pool DB queries.

Calls to `atomically` cannot be nested.

jonathanknowles committed:
Rename to `PoolRegistrationStatus` to `PoolLifeCycleStatus`.

Justification:

This type is aimed at capturing the different stages in a pool's lifecycle.

The name `LifeCycleStatus` avoids overt association with any single
aspect of the life cycle of a pool, such as registration or retirement.

jonathanknowles committed:
Increase the integration test timeout to 5 minutes.
